csharp/ql/src/experimental/Security Features/campaign/Solorigate/ModifiedFnvFunctionDetection.ql

```diff
@@ -0,0 +1,32 @@
+/**
+ * @name Detects a modified FNV function
+ * @description Possible indication of Solorigate. Detects a modified FNV1 function, where there is an additional xor using a literal after the regular FNV hash
+ * @kind problem
+ * @tags security
+ *       solorigate
+ * @precision high
+ * @id cs/solorigate/modified-fnv-function-detection
+ * @problem.severity error
+ */
+
+import csharp
+import Solorigate
+import microsoft.code.csharp.Cryptography.NonCryptographicHashes
+
+from Variable v, Literal l, LoopStmt loop, Expr additional_xor
+where
+  maybeUsedInFNVFunction(v, _, _, loop) and
+  (
+    exists(BitwiseXorExpr xor2 | xor2.getAnOperand() = l and additional_xor = xor2 |
+      loop.getAControlFlowNode().getASuccessor*() = xor2.getAControlFlowNode() and
+      xor2.getAnOperand() = v.getAnAccess()
+    )
+    or
+    exists(AssignXorExpr xor2 | xor2.getAnOperand() = l and additional_xor = xor2 |
+      loop.getAControlFlowNode().getASuccessor*() = xor2.getAControlFlowNode() and
+      xor2.getAnOperand() = v.getAnAccess()
+    )
+  )
```

Comment:
Are you checking that the additional XOR happens _after_ the loop or after the multiplication detected by `maybeUsedInFNVFunction` _inside_ the loop?

```csharp
while (...) {
  x ^= y;
  x *= z;
}
x ^= tamper;
```

vs.

```csharp
while (...) {
  x ^= y;
  x *= z;
  x ^= tamper;
}
```

In the former case, `getAControlFlowNode()` should be replaced with `getAControlFlowExitNode()` instead; in the latter case we would instead have to check that it can be reached from the multiplication, and that it can reach back to the first XOR (i.e., that it is still inside the loop).

Reply:
We want the former, so I am fixing it with getAControlFlowExitNode. Thanks